drivers: usb: udc_stm32: Handle HAL callbacks on separate thread

topisani opened the pull request:
All HAL callback handling is offloaded to a separate thread.
Tested on STM32H747 (arduino portenta) with the high speed controller.

This is an attempt at fixing https://github.com/zephyrproject-rtos/zephyr/issues/61464.
Marked as a draft, as i would at least like to do some minor stylistic cleanup before submitting, but if anyone could help test, I would appreciate it. As far as i can tell, no additional locking should be required, and no timing invariants are broken, but I could use some extra eyes on that as well.

Fixes: #61464 (UDC STM32 part)

topisani committed:
drivers: udc: stm32: Offload data callbacks to thread

All HAL callback handling is offloaded to a separate thread, as they
involve non isr-compatible operations such as mutexes.

Fixes #61464

Signed-off-by: Tobias Pisani <mail@topisani.dev>

erwango commented:
A huge thanks. We'll for sure help testing on extended sample of series.

GeorgeCGV commented:
Is it safe to enqueue a stack-allocated msg pointer?

Perhaps the event approach is safer. Given that the combined number of IN and OUT endpoints is typically less than 32 (_15 IN + 15 OUT_), a `k_event` could be used to notify about IN, OUT, or SETUP.

Additionally, this approach helps prevent potential message queue overflows that might result in missed event/message handling and supports ordered processing.

topisani commented:
> Is it safe to enqueue a stack-allocated msg pointer?

Yes, it is copied into the ringbuffer.

> Perhaps the event approach is safer. Given that the combined number of IN and OUT endpoints is typically less than 32 (_15 IN + 15 OUT_), a `k_event` could be used to notify about IN, OUT, or SETUP.

Isnt it technically up to 16+16 endpoints including the control endpoint?

> Additionally, this approach helps prevent potential message queue overflows that might result in missed event/message handling and supports ordered processing.

While I agree about the message queue overflows, I don't see how k_event would support ordered processing? Thats mostly why I chose the msgq based approach, to make sure the events were processed in the order they arrived

topisani commented:
> Not in a fifo way, but In the processing. I.e. process `setup` before `in` (_process `x` before `y`_).

Ah, I dont think this is what we would want though. All of these callbacks are triggered from the same IRQ, i.e. have the same priority, and should be processed in the order they arrived. It should maybe be considered how the ordering interdependency with the `udc_submit_event` processing could impact things, but i think that requires a deeper knowledge of the STM32 USB stack than I have. I would personally be ok to leave this up to testing too.

@erwango I will rebase this, but then I believe it is ready for testing. Is there anything else I can do to help the testing effort along?

erwango commented:
@topisani It's failing on `disco_l475_iot1`. Before:
```
*** Booting Zephyr OS build v4.0.0-405-g2efc8598e39f ***
[00:00:00.010,000] <inf> cdc_acm_echo: USB device support enabled
[00:00:00.010,000] <inf> cdc_acm_echo: Wait for DTR
[00:00:25.413,000] <inf> cdc_acm_echo: USBD message: Device suspended
[00:00:25.568,000] <inf> cdc_acm_echo: USBD message: Bus reset
[00:00:25.683,000] <inf> cdc_acm_echo: USBD message: Bus reset
[00:00:25.760,000] <inf> cdc_acm_echo: USBD message: New device configuration
[00:00:25.761,000] <err> udc_stm32: ep 0x00 queue is empty
[00:00:25.761,000] <inf> cdc_acm_echo: USBD message: CDC ACM line coding
[00:00:25.761,000] <inf> cdc_acm_echo: Baudrate 9600
```
After:
```
*** Booting Zephyr OS build v4.0.0-409-g84574a1c389f ***
[00:00:00.010,000] <err> usbd_init: Failed to assign endpoint addresses
[00:00:00.010,000] <err> usbd_init: Failed to init HS configuration 1
[00:00:00.010,000] <err> usbd_sample_config: Failed to initialize device support
[00:00:00.010,000] <err> cdc_acm_echo: Failed to initialize USB device
[00:00:00.010,000] <err> cdc_acm_echo: Failed to enable USB
```

topisani commented:
@erwango Alright, i believe this is my attempt at detecting USB HS support that fails. I have an l432 board i can test on, if that is the case, it should fail there as well

topisani committed:
drivers: udc: stm32: set the hs capability if high speed is available

This fixes usbd_caps_speed(), which is used in sample_usbd_init, and the
documentation. Without it, no high speed configuration would be added.

Signed-off-by: Tobias Pisani <mail@topisani.dev>

topisani commented:
I could reproduce the issue on a nucleo_l432kc, and it appears to have been the HS detection. It now follows the maximum-speed dts prop, like in most of the other udc drivers.

topisani commented:
Alright, i did some testing using `samples/subsys/usb/cdc_acm -DCONF_FILE=usbd_next_prj.conf` on the various boards i could find.
test: open the cdc_acm terminal, and type characteres. The device should echo each character back.

`nucleo_u575zi_q`: PASS
`stm32f469i_disco`: PASS
`stm32f769i_disco`: FAIL

I have also done a test on an arduino_portenta_h7 based custom board, that also failed. The common factor is that both boards use HS USB.

Checking with `main`, these boards also fail, so I think the current state is just that udc_stm32 does not correctly implement usbotg_hs. I will look at a fix separately from this PR.

fpistm commented:
Hi @topisani 

I've started to test on my side, here my result using `samples/subsys/usb/cdc_acm`  with `-DCONF_FILE=usbd_next_prj.conf` and sent a lorem ipsum buffer of 1024 bytes several times.
| series| boards | controller | cdc_acm sample | plug/unplug |
| --- | --- | --- | --- | --- |
| F4 | nucleo_f429zi | st,stm32-otgfs | :white_check_mark: | :white_check_mark: |
| F7 | stm32f746g_disco | st,stm32-otgfs | :white_check_mark: | :white_check_mark: |
| H7 | nucleo_h723zg | st,stm32-otgfs | :white_check_mark: | :white_check_mark: |
| L4 | b_l4s5i_iot01a | st,stm32-otgfs | :white_check_mark: | :white_check_mark: |

I will continue tests and report here.

fpistm commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -912,6 +913,7 @@ static const struct udc_stm32_config udc0_cfg = {
 	.pma_offset = USB_BTABLE_SIZE,
 	.ep0_mps = EP0_MPS,
 	.ep_mps = EP_MPS,
+	.speed_idx = DT_ENUM_IDX(DT_DRV_INST(0), maximum_speed),
```
maximum_speed property is not available from all DT.
In old driver, property is check before access it:
https://github.com/zephyrproject-rtos/zephyr/blob/c0be51df2fb83f2e772d74b73e10350662af12b8/drivers/usb/device/usb_dc_stm32.c#L70-L72

erwango replied:
```suggestion
	.speed_idx = DT_ENUM_IDX_OR(DT_DRV_INST(0), maximum_speed, low-speed),
```

topisani commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -260,6 +243,28 @@ void HAL_PCD_DataOutStageCallback(PCD_HandleTypeDef *hpcd, uint8_t epnum)
 {
 	uint32_t rx_count = HAL_PCD_EP_GetRxCount(hpcd, epnum);
 	struct udc_stm32_data *priv = hpcd2data(hpcd);
+	struct udc_stm32_msg msg = {
```
I caught an issue where this check would trigger `ZLP expected` (only appeared on usbotg_hs), as multiple messages arriving before being handled would be merged. Storing the rx length in the msgq fixed this. It doubles the size of the messages, but I dont see a way around it. The cdc_acm test still fails on usbotg_hs.
https://github.com/zephyrproject-rtos/zephyr/blob/c0be51df2fb83f2e772d74b73e10350662af12b8/drivers/usb/udc/udc_common.c#L1058-L1064

erwango replied:
Nice! I think this could potentially fix a couple of issues seen by @fpistm (particularly on Windows). He'll only be able to answer on monday though.

> The cdc_acm test still fails on usbotg_hs.

Which board are you using for this test ?

topisani replied:
> Which board are you using for this test ?
Mainly testing using stm32f769i_disco, but i see the same on a portenta_h7. I may be able to test the fs interface on the h7 to verify that it only happens on hs, but the portenta is mounted in a custom board where its a bit more difficult to perform isolated tests.
Just to clarify: when running the cdc_acm sample, descriptors are exchanged without issue, but no characters are echoed back. I am currently doing some debugging with wireshark, i will report a separate issue once i have more context.

marwaiehm-st replied:
@topisani 
I experienced the same problem of characters not being echoed back with the stm32h747i_disco board, but it ran smoothly on the nucleo_h743zi board, which means it only occurs on HS.

topisani replied:
i have now also confirmed that it works on the usbotg_fs interface on stm32f769i_disco. This issue with HS also exists on main.

topisani committed:
drivers: udc: stm32: Fix RX FIFO min size

On USB HS, the previous lower limit of 64 is too small, the value
160 has been chosen apparently through trial and error.

See 1204aa25 for the original implementation in the old device driver.

Signed-off-by: Tobias Pisani <mail@topisani.dev>

topisani commented:
I found the usbotg_hs issue - porting this fix to udc helped: https://github.com/zephyrproject-rtos/zephyr/pull/62530
Like the original related issues and PRs, i cannot find an explanation as to why 160 works as a minimum, so i kept the comment that it was chosen through trial and error.

dsch commented:
Thank you @topisani for your great work!

Is it also required to push request to `udc_ep_enqueue()` on the message queue?

Similar as in NRF driver: https://github.com/zephyrproject-rtos/zephyr/blob/6dafb5f4a9cbd7043989f986d1f5963c264d413e/drivers/usb/udc/udc_nrf.c#L584-L604

topisani commented on drivers/usb/udc/Kconfig.stm32:
```diff
@@ -6,6 +6,7 @@ config UDC_STM32
 	depends on DT_HAS_ST_STM32_OTGFS_ENABLED \
 		|| DT_HAS_ST_STM32_OTGHS_ENABLED \
 		|| DT_HAS_ST_STM32_USB_ENABLED
+	depends on !USB_DC_STM32
```
It's also a bit specific to this driver though - at least theoretically it should be possible with multi-instance compatible controllers to use one with the new api, and one with the old. But the two stm32 drivers specifically cannot coexist because of the Hal callback definitions.

erwango replied:
ok, Thanks. Fine for me then

topisani replied:
ok - would you recomend removing this check all together, or moving it into an #ifdef in the driver?

topisani commented:
> Is it also required to push request to `udc_ep_enqueue()` on the message queue?

I considered it, but as far as I can tell it should be ok. I will take another look

jfischer-no commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -515,9 +515,9 @@ static void udc_stm32_mem_init(const struct device *dev)
+	 * 160 has been selected through trial and error.
 	 */
-	words = MAX(0x40, cfg->ep_mps / 4);
+	words = MAX(160, cfg->ep_mps / 4);
```
>On USB HS, the previous lower limit of 64 is too small, the value
160 has been chosen apparently through trial and error.

At a minimum, this value should be selected based on the controller type and capabilities, not "trial and error".

topisani replied:
See the discussion here: https://github.com/zephyrproject-rtos/zephyr/pull/62530
I've been digging through documentation and related issues from when this change was introduced on the old usb device driver, and i cannot seem to find any reason. This configuration is what worked in the old driver, which has been used a lot in practice. But maybe someone at ST can help settle where this value comes from (@erwango)?

erwango replied:
Unfortunately @Desvauxm-st left the company. So I don't have more information other than what is available in the PR. @marwaiehm-st maybe ?

marwaiehm-st replied:
Unfortunately, we tried to find a satisfactory explanation other than "this value works", but until now, we haven't found an explanation for why the minimum value of FIFO RX needs to be 160.

jfischer-no replied:
I am okay with this value. But, "Calculating FIFO Size" is described in DWC2 Programming Guide, you should have access to this documentation internally. IIRC the FIFO calculation for the one other type is much simpler. Instead of continuing "trial and error" with this driver, STM32 controller should start using DWC2 controller and clean up this driver to be used with non-DWC2 type controller.

topisani replied:
> STM32 controller should start using DWC2 controller and clean up this driver to be used with non-DWC2 type controller.

Is this a discussion we can track in a separate issue? Some consideration was made years ago (https://github.com/zephyrproject-rtos/zephyr/issues/4412), but it seems like an area where a decision should be made, and this PR seems like the wrong place to have that discussion.

jfischer-no replied:
No, this is the right place to make a comment about it and it is related to FIFO size calculation.

topisani commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -123,6 +135,24 @@ void HAL_PCD_ResumeCallback(PCD_HandleTypeDef *hpcd)
 	udc_submit_event(priv->dev, UDC_EVT_RESUME, 0);
 }
 
+void HAL_PCD_SetupStageCallback(PCD_HandleTypeDef *hpcd)
+{
+	struct udc_stm32_data *priv = hpcd2data(hpcd);
+	struct udc_stm32_msg msg = {.kind = UDC_STM32_MSG_SETUP};
+	int err = k_msgq_put(&priv->msgq_data, &msg, K_NO_WAIT);
+
+	if (err < 0) {
+		LOG_ERR("UDC Message queue overrun");
+	}
```
The driver specifically needs to handle separate RX callbacks individually. I dont see a way to do this using events, as multiple callbacks for one endpoint can occur between the handler thread running. I assume this is the reason why the udc_nrf driver also does it this way.

jfischer-no replied:
No, nRF USBD driver does it this way because it was introduced before @tmon-nordic  figured out a better way to handle it in DWC2 driver.

topisani replied:
I can take a look at reimplementing it with that approach, if it is considered a blocker to getting this bug fixed. That means the development and testing effort already put into this issue have to be redone. Personally, I would prefer to get this bug fixed with an approach that is clearly acceptible, as it is still being used in udc_nrf, and also in a driver that in your opinion should be heavily reworked anyway.

jfischer-no replied:
You do not need to change it here, but it is my duty to point out a possible better solution.

fpistm commented:
Hi,
here the result of all my tests using `samples/subsys/usb/cdc_acm`  with `-DCONF_FILE=usbd_next_prj.conf` and sent a lorem ipsum buffer of 1024 bytes several times.

`In any case it fixes the assert issue reported here: #61464.`

| series| boards | controller | cdc_acm sample | plug/unplug | comment |
| --- | --- | --- | --- | --- | --- |
| F2 | nucleo_f207zg | st,stm32-otgfs | :green_circle: | :green_circle: | |
| F4 | nucleo_f429zi | st,stm32-otgfs | :green_circle: | :green_circle: | |
| F7 | stm32f746g_disco | st,stm32-otgfs | :green_circle: | :green_circle: | |
| H7 | nucleo_h723zg | st,stm32-otghs | :green_circle: | :green_circle: | |
| L4 | b_l4s5i_iot01a | st,stm32-otgfs | :green_circle: | :green_circle: | |
| U5 | nucleo_u575zi_q | st,stm32-otgfs | :green_circle: | :green_circle: | |
| F3 | nucleo_f303re (1)| st,stm32-usb | :orange_circle: | :red_circle: | [dts updated](#dts-updated). See [here](#nucleo_f303re) |
| G4 | nucleo_g431rb (1)| st,stm32-usb | :orange_circle: | :red_circle: | [dts updated](#dts-updated). Same status than [nucleo_f303re](#nucleo_f303re) except no patch needed to get a fixed baudrate value (115200) |
| H5 | nucleo_h563zi | st,stm32-usb | :orange_circle: using [patch](#workaround-1) | :red_circle: | Bus fault #76250. Fixed baudrate value reported (115200) using [patch](#workaround-1) |
| WB | nucleo_wb55rg | st,stm32-usb | :orange_circle: | :red_circle: | Same status than [nucleo_f303re](#nucleo_f303re). |

> [!NOTE]
> 1. shield used to add USB connector

##### Workaround 1
  * #76250 `subsys/usb/device_next/class/usbd_cdc_acm.c`
```patch
@@ -458,11 +458,14 @@ static int usbd_cdc_acm_ctd(struct usbd_class_data *const c_data,
 		if (setup->wLength != len) {
 			errno = -ENOTSUP;
 			return 0;
 		}

-		memcpy(&data->line_coding, buf->data, len);
+		if (buf->data) {
+			printk("SET_LINE_CODING buf->data not null\n");
+			memcpy(&data->line_coding, buf->data, len);
+		}
 		cdc_acm_update_uart_cfg(data);
 		usbd_msg_pub_device(uds_ctx, USBD_MSG_CDC_ACM_LINE_CODING, dev);
 		return 0;

 	case SET_CONTROL_LINE_STATE:
  ```

##### dts updated
```patch
+
+zephyr_udc0: &usb {
+	pinctrl-0 = <&usb_dm_pa11 &usb_dp_pa12>;
+	pinctrl-names = "default";
+	status = "okay";
+};
```

##### nucleo_f303re
 * Wrong Baudrate value reported: `<inf> cdc_acm_echo: Baudrate 536885760`.
 * With [patch](#workaround-1) value is fixed (115200).
   Seems same issue than #76250, no crash because the `buf->data` has a value but not a correct one.
   Old stack also report a fixed baudrate.
 * [Assertion](#assert-enabled) if `CONFIG_ASSERT=y`
 * [Plug/unplug](#plugunplug) issue


##### Assert enabled

<details>
  <summary>log</summary>

```log
*** Booting Zephyr OS build v4.0.0-1636-gddeb51a89cfd ***
[00:00:00.001,000] <err> usbd_core: Failed to set default address after bus reset:
[00:00:00.001,000] <inf> cdc_acm_echo: USBD message: Bus reset:
[00:00:00.001,000] <inf> cdc_acm_echo: USBD message: Stack error:
[00:00:00.001,000] <inf> cdc_acm_echo: USB device support enabled:
[00:00:00.001,000] <inf> cdc_acm_echo: Wait for DTR:
[00:00:00.004,000] <inf> cdc_acm_echo: USBD message: Device suspended:
[00:00:00.112,000] <inf> cdc_acm_echo: USBD message: Device resumed:
[00:00:00.167,000] <inf> cdc_acm_echo: USBD message: Bus reset:
[00:00:00.281,000] <inf> cdc_acm_echo: USBD message: Bus reset:
[00:00:00.362,000] <inf> cdc_acm_echo: USBD message: New device configuration:
ASSERTION FAIL [net_buf_simple_tailroom(buf) >= len] @ WEST_TOPDIR/zephyr/lib/net_buf/buf_simple.c:62
[00:00:00.363,000] <err> os: r0/a1:  0x00000004  r1/a2:  0x0000003e  r2/a3:  0x00000005:
[00:00:00.363,000] <err> os: r3/a4:  0x00000004 r12/ip:  0x20003160 r14/lr:  0x08003b5d:
[00:00:00.363,000] <err> os:  xpsr:  0x01000000:
[00:00:00.363,000] <err> os: Faulting instruction address (r15/pc): 0x0800d228:
[00:00:00.363,000] <err> os: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0:
[00:00:00.363,000] <err> os: Current thread: 0x20000f60 (unknown):
[00:00:00.444,000] <err> os: Halting system:
```
</details>

#### Plug/unplug
> [!IMPORTANT]
> On linux, second plug produce this issue (re-enumeration):
> On Windows, first plug (enumeration), means CDC port never created.

<details>
  <summary>log</summary>

```log
[00:00:07.187,000] <err> udc: Failed to allocate net_buf 2
[00:00:07.187,000] <err> usbd_ch9: Buffer for data|status is missing
[00:00:07.187,000] <err> usbd_ch9: Malformed setup packet
[00:00:07.188,000] <err> udc: Failed to allocate net_buf 2
[00:00:07.188,000] <err> usbd_ch9: Buffer for data|status is missing
[00:00:07.188,000] <err> usbd_ch9: Malformed setup packet
[00:00:07.188,000] <err> udc: Failed to allocate net_buf 2
[00:00:07.188,000] <err> usbd_ch9: Buffer for data|status is missing
[00:00:07.188,000] <err> usbd_ch9: Malformed setup packet
[00:00:07.188,000] <err> udc: Failed to allocate net_buf 255
[00:00:07.188,000] <err> usbd_ch9: Buffer for data|status is missing
[00:00:07.188,000] <err> usbd_ch9: Malformed setup packet
[00:00:07.188,000] <err> udc: Failed to allocate net_buf 255
[00:00:07.188,000] <err> usbd_ch9: Buffer for data|status is missing
[00:00:07.188,000] <err> usbd_ch9: Malformed setup packet
[00:00:07.189,000] <err> udc: Failed to allocate net_buf 255
[00:00:07.189,000] <err> usbd_ch9: Buffer for data|status is missing
[00:00:07.189,000] <err> usbd_ch9: Malformed setup packet
[00:00:07.189,000] <err> udc: Failed to allocate net_buf 2
[00:00:07.189,000] <err> usbd_ch9: Buffer for data|status is missing
[00:00:07.189,000] <err> usbd_ch9: Malformed setup packet
[00:00:07.189,000] <err> udc: Failed to allocate net_buf 2
[00:00:07.189,000] <err> usbd_ch9: Buffer for data|status is missing
[00:00:07.189,000] <err> usbd_ch9: Malformed setup packet
[00:00:07.190,000] <err> udc: Failed to allocate net_buf 2
[00:00:07.190,000] <err> usbd_ch9: Buffer for data|status is missing
[00:00:07.190,000] <err> usbd_ch9: Malformed setup packet
[00:00:07.193,000] <err> udc: Failed to allocate net_buf 0
[00:00:07.193,000] <err> usbd_ch9: Buffer for data|status is missing
[00:00:07.193,000] <err> usbd_ch9: Malformed setup packet
```

</details>

erwango commented:
> here the result of all my tests using samples/subsys/usb/cdc_acm with -DCONF_FILE=usbd_next_prj.conf and sent a lorem ipsum buffer of 1024 bytes several times.

Thanks @fpistm. Hence to summarize, this PR actually fixes https://github.com/zephyrproject-rtos/zephyr/issues/61464, but some issues remain, likely all linked/close to #76250.

For clarity, I'll open a new issue to gather all this.

topisani commented:
Addressed most review comments from @jfischer-no. For the remaining ones, I still need some clarification on how to proceed.

erwango commented:
@jfischer-no PTAL

jfischer-no commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -259,6 +239,28 @@ void HAL_PCD_DataOutStageCallback(PCD_HandleTypeDef *hpcd, uint8_t epnum)
+{
+	struct udc_stm32_data *priv = hpcd2data(hpcd);
+	struct udc_stm32_msg msg = {.kind = UDC_STM32_MSG_DATA_IN, .ep = epnum};
+	int err = k_msgq_put(&priv->msgq_data, &msg, K_NO_WAIT);
+
+	if (err < 0) {
```
Not addressed.

jfischer-no commented on drivers/usb/udc/Kconfig.stm32:
```diff
@@ -13,3 +13,25 @@ config UDC_STM32
 	default y
 	help
 	  STM32 USB device controller driver.
+
+
+config UDC_STM32_STACK_SIZE
+	int "UDC controller driver internal thread stack size"
+	depends on UDC_STM32
+	default 512
+	help
+	  STM32 USB device controller driver internal thread stack size.
+
+config UDC_STM32_THREAD_PRIORITY
+	int "STM32 USB controller driver thread priority"
+	depends on UDC_STM32
+	default 8
+	help
+	  STM32 USB device controller driver thread priority.
+
+config UDC_STM32_MAX_QMESSAGES
+	int "STM32 UDC driver maximum number of ISR event messages"
+	range 4 64
+	default 8
+	help
+	  Maximum number of messages for handling of STM32 USBD ISR events.
```
This pollutes the list.

jfischer-no replied:
That should be

```Kconfig
	  STM32 USB device controller driver.

if UDC_STM32

config UDC_STM32_STACK_SIZE
	int "UDC controller driver internal thread stack size"
	default 512
	help
	  STM32 USB device controller driver internal thread stack size.

config UDC_STM32_THREAD_PRIORITY
	int "STM32 USB controller driver thread priority"
	default 8
	help
	  STM32 USB device controller driver thread priority.

config UDC_STM32_MAX_QMESSAGES
	int "STM32 UDC driver maximum number of ISR event messages"
	range 4 64
	default 8
	help
	  Maximum number of messages for handling of STM32 USBD ISR events.

endif
```

btw, this is quite trivial and easy to spot.

jfischer-no commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -143,59 +175,7 @@ static int usbd_ctrl_feed_dout(const struct device *dev, const size_t length)
-
-static int udc_stm32_tx(const struct device *dev, uint8_t ep,
-			struct net_buf *buf)
+static int udc_stm32_tx(const struct device *dev, uint8_t ep, struct net_buf *buf)
```
Stray change.

jfischer-no commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -263,6 +243,28 @@ void HAL_PCD_DataOutStageCallback(PCD_HandleTypeDef *hpcd, uint8_t epnum)
 {
 	uint32_t rx_count = HAL_PCD_EP_GetRxCount(hpcd, epnum);
 	struct udc_stm32_data *priv = hpcd2data(hpcd);
+	struct udc_stm32_msg msg = {
+		.type = UDC_STM32_MSG_DATA_OUT, .ep = epnum, .rx_count = rx_count};
+	int err = k_msgq_put(&priv->msgq_data, &msg, K_NO_WAIT);
+
+	if (err < 0) {
+		LOG_ERR("UDC Message queue overrun");
+	}
```
Use the usual style, use the usual style in UDC drivers, keep the style in this file. Do not hide the calls (they are not helpers like udc_get_private()). Do not introduce undefined behavior. Fix it everywhere.
```c
	struct udc_stm32_msg msg = {
		.type = UDC_STM32_MSG_DATA_OUT,
		.ep = epnum,
		.rx_count = rx_count,
	};
	int err;

	err = k_msgq_put(&priv->msgq_data, &msg, K_NO_WAIT);
	if (err != 0) {
		LOG_ERR("UDC Message queue overrun");
	}
```

jfischer-no commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -263,6 +243,28 @@ void HAL_PCD_DataOutStageCallback(PCD_HandleTypeDef *hpcd, uint8_t epnum)
+	}
+}
+
+static void priv_handle_msg_data_out(struct udc_stm32_data *priv, uint8_t epnum, uint16_t rx_count)
```
What does `priv_` mean? It is static, it does not make it more private if it is prefixed with `priv_`. It is just confusing, please remove it.

erwango commented:
@topisani Please rebase.

jfischer-no commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -275,6 +256,33 @@ void HAL_PCD_DataOutStageCallback(PCD_HandleTypeDef *hpcd, uint8_t epnum)
+void HAL_PCD_DataInStageCallback(PCD_HandleTypeDef *hpcd, uint8_t epnum)
+{
+	struct udc_stm32_data *priv = hpcd2data(hpcd);
+	struct udc_stm32_msg msg = {.type = UDC_STM32_MSG_DATA_IN, .ep = epnum};
+	int err;
```
Still not fixed
```c
	struct udc_stm32_msg msg = {
		.type = UDC_STM32_MSG_DATA_IN,
		.ep = epnum
	};
	int err;
```

ddavidebor commented:
A big thank you for this PR!
